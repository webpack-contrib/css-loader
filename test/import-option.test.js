const { webpack, evaluated, normalizeErrors } = require('./helpers');

describe('import option', () => {
  it('true', async () => {
    const testId = './import/import.css';
    const stats = await webpack(testId);
    const { modules } = stats.toJson();
    const module = modules.find((m) => m.id === testId);

    expect(module.source).toMatchSnapshot('module');
    expect(evaluated(module.source, modules)).toMatchSnapshot(
      'module (evaluated)'
    );
    expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
      'warnings'
    );
    expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot('errors');
  });

  it('false', async () => {
    const config = { loader: { options: { import: false } } };
    const testId = './import/import.css';
    const stats = await webpack(testId, config);
    const { modules } = stats.toJson();
    const module = modules.find((m) => m.id === testId);

    expect(module.source).toMatchSnapshot('module');
    expect(evaluated(module.source, modules)).toMatchSnapshot(
      'module (evaluated)'
    );
    expect(stats.compilation.warnings).toMatchSnapshot('warnings');
    expect(stats.compilation.errors).toMatchSnapshot('errors');
  });

  [true, false].forEach((modulesValue) => {
    it(`false and modules ${modulesValue}`, async () => {
      const config = {
        loader: { options: { import: false, modules: modulesValue } },
      };
      const testId = './import/css-modules.css';
      const stats = await webpack(testId, config);
      const { modules } = stats.toJson();
      const module = modules.find((m) => m.id === testId);

      expect(module.source).toMatchSnapshot('module');
      expect(evaluated(module.source, modules)).toMatchSnapshot(
        'module (evaluated)'
      );
      expect(normalizeErrors(stats.compilation.warnings)).toMatchSnapshot(
        'warnings'
      );
      expect(normalizeErrors(stats.compilation.errors)).toMatchSnapshot(
        'errors'
      );
    });
  });
});
