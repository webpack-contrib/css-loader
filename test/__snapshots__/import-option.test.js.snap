// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import option false and modules false: errors 1`] = `Array []`;

exports[`import option false and modules false: module (evaluated) 1`] = `
Array [
  Array [
    2,
    "
",
    "",
  ],
  Array [
    1,
    "@import url(test-other.css) (min-width: 100px);

.ghi {
  color: red;
}
",
    "",
  ],
]
`;

exports[`import option false and modules false: module 1`] = `
"exports = module.exports = require(\\"../../../lib/runtime/api.js\\")(false);
// imports
exports.i(require(\\"-!../../../index.js??ref--4-0!./values.css\\"), \\"\\");

// module
exports.push([module.id, \\"@import url(test-other.css) (min-width: 100px);\\\\n\\\\n.ghi {\\\\n  color: \\" + require(\\"-!../../../index.js??ref--4-0!./values.css\\").locals[\\"def\\"] + \\";\\\\n}\\\\n\\", \\"\\"]);

// exports
exports.locals = {
	\\"def\\": \\"\\" + require(\\"-!../../../index.js??ref--4-0!./values.css\\").locals[\\"def\\"] + \\"\\"
};"
`;

exports[`import option false and modules false: warnings 1`] = `Array []`;

exports[`import option false and modules true: errors 1`] = `Array []`;

exports[`import option false and modules true: module (evaluated) 1`] = `
Array [
  Array [
    2,
    "
",
    "",
  ],
  Array [
    1,
    "@import url(test-other.css) (min-width: 100px);

._3r49KZIIAltPknAjdNVZ-7 {
  color: red;
}
",
    "",
  ],
]
`;

exports[`import option false and modules true: module 1`] = `
"exports = module.exports = require(\\"../../../lib/runtime/api.js\\")(false);
// imports
exports.i(require(\\"-!../../../index.js??ref--4-0!./values.css\\"), \\"\\");

// module
exports.push([module.id, \\"@import url(test-other.css) (min-width: 100px);\\\\n\\\\n._3r49KZIIAltPknAjdNVZ-7 {\\\\n  color: \\" + require(\\"-!../../../index.js??ref--4-0!./values.css\\").locals[\\"def\\"] + \\";\\\\n}\\\\n\\", \\"\\"]);

// exports
exports.locals = {
	\\"def\\": \\"\\" + require(\\"-!../../../index.js??ref--4-0!./values.css\\").locals[\\"def\\"] + \\"\\",
	\\"ghi\\": \\"_3r49KZIIAltPknAjdNVZ-7\\"
};"
`;

exports[`import option false and modules true: warnings 1`] = `Array []`;

exports[`import option false: errors 1`] = `Array []`;

exports[`import option false: module (evaluated) 1`] = `
Array [
  Array [
    1,
    "@import url(test.css);
@import url('test.css');
@import url(\\"test.css\\");
@IMPORT url(test.css);
@import URL(test.css);
@import url(test.css );
@import url( test.css);
@import url( test.css );
@import url(
  test.css
);
@import url();
@import url('');
@import url(\\"\\");
@import \\"test.css\\";
@import 'test.css';
@import '';
@import \\"\\";
@import \\"   \\";
@import \\"
\\";
@import url();
@import url('');
@import url(\\"\\");
@import url(test.css) screen and print;
@import url(test.css) SCREEN AND PRINT;
@import url(test.css)screen and print;
@import url(test.css)  screen and print;
@import url(test-media.css) screen and print;
@import url(test-other.css) (min-width: 100px);
@import url(http://example.com/style.css);
@import url(http://example.com/style.css);
@import url(http://example.com/style.css#hash);
@import url(http://example.com/style.css?#hash);
@import url(http://example.com/style.css?foo=bar#hash);
@import url(http://example.com/other-style.css) screen and print;
@import url(http://example.com/other-style.css) screen and print;
@import url(\\"//example.com/style.css\\");
@import url(~package/test.css);
@import ;
@import foo-bar;
@import-normalize;
@import url('http://') :root {}
@import url('query.css?foo=1&bar=1');
@import url('other-query.css?foo=1&bar=1#hash');
@import url('other-query.css?foo=1&bar=1#hash') screen and print;
@import url('https://fonts.googleapis.com/css?family=Roboto');
@import url('https://fonts.googleapis.com/css?family=Noto+Sans+TC');
@import url('https://fonts.googleapis.com/css?family=Noto+Sans+TC|Roboto');

.class {
  a: b c d;
}

.foo {
  @import 'path.css';
}
",
    "",
  ],
]
`;

exports[`import option false: module 1`] = `
"exports = module.exports = require(\\"../../../lib/runtime/api.js\\")(false);
// imports


// module
exports.push([module.id, \\"@import url(test.css);\\\\n@import url('test.css');\\\\n@import url(\\\\\\"test.css\\\\\\");\\\\n@IMPORT url(test.css);\\\\n@import URL(test.css);\\\\n@import url(test.css );\\\\n@import url( test.css);\\\\n@import url( test.css );\\\\n@import url(\\\\n  test.css\\\\n);\\\\n@import url();\\\\n@import url('');\\\\n@import url(\\\\\\"\\\\\\");\\\\n@import \\\\\\"test.css\\\\\\";\\\\n@import 'test.css';\\\\n@import '';\\\\n@import \\\\\\"\\\\\\";\\\\n@import \\\\\\"   \\\\\\";\\\\n@import \\\\\\"\\\\n\\\\\\";\\\\n@import url();\\\\n@import url('');\\\\n@import url(\\\\\\"\\\\\\");\\\\n@import url(test.css) screen and print;\\\\n@import url(test.css) SCREEN AND PRINT;\\\\n@import url(test.css)screen and print;\\\\n@import url(test.css)  screen and print;\\\\n@import url(test-media.css) screen and print;\\\\n@import url(test-other.css) (min-width: 100px);\\\\n@import url(http://example.com/style.css);\\\\n@import url(http://example.com/style.css);\\\\n@import url(http://example.com/style.css#hash);\\\\n@import url(http://example.com/style.css?#hash);\\\\n@import url(http://example.com/style.css?foo=bar#hash);\\\\n@import url(http://example.com/other-style.css) screen and print;\\\\n@import url(http://example.com/other-style.css) screen and print;\\\\n@import url(\\\\\\"//example.com/style.css\\\\\\");\\\\n@import url(~package/test.css);\\\\n@import ;\\\\n@import foo-bar;\\\\n@import-normalize;\\\\n@import url('http://') :root {}\\\\n@import url('query.css?foo=1&bar=1');\\\\n@import url('other-query.css?foo=1&bar=1#hash');\\\\n@import url('other-query.css?foo=1&bar=1#hash') screen and print;\\\\n@import url('https://fonts.googleapis.com/css?family=Roboto');\\\\n@import url('https://fonts.googleapis.com/css?family=Noto+Sans+TC');\\\\n@import url('https://fonts.googleapis.com/css?family=Noto+Sans+TC|Roboto');\\\\n\\\\n.class {\\\\n  a: b c d;\\\\n}\\\\n\\\\n.foo {\\\\n  @import 'path.css';\\\\n}\\\\n\\", \\"\\"]);

// exports
"
`;

exports[`import option false: warnings 1`] = `Array []`;

exports[`import option true: errors 1`] = `Array []`;

exports[`import option true: module (evaluated) 1`] = `
Array [
  Array [
    2,
    ".test {
  a: a;
}
",
    "",
  ],
  Array [
    3,
    ".test {
  a: a;
}
",
    "screen and print",
  ],
  Array [
    5,
    "a {
  b: b;
}
",
    "((min-width: 100px)) and (screen and print)",
  ],
  Array [
    4,
    ".test {
  c: c;
}
",
    "screen and print",
  ],
  Array [
    1,
    "@import url(http://example.com/style.css);",
    "",
  ],
  Array [
    1,
    "@import url(http://example.com/style.css#hash);",
    "",
  ],
  Array [
    1,
    "@import url(http://example.com/style.css?#hash);",
    "",
  ],
  Array [
    1,
    "@import url(http://example.com/style.css?foo=bar#hash);",
    "",
  ],
  Array [
    1,
    "@import url(http://example.com/other-style.css);",
    "screen and print",
  ],
  Array [
    1,
    "@import url(//example.com/style.css);",
    "",
  ],
  Array [
    6,
    ".test {
  d: d
}
",
    "",
  ],
  Array [
    7,
    ".query {
  e: e;
}
",
    "",
  ],
  Array [
    8,
    ".other-query {
  f: f;
}
",
    "",
  ],
  Array [
    9,
    ".other-query {
  f: f;
}
",
    "screen and print",
  ],
  Array [
    1,
    "@import url(https://fonts.googleapis.com/css?family=Roboto);",
    "",
  ],
  Array [
    1,
    "@import url(https://fonts.googleapis.com/css?family=Noto+Sans+TC);",
    "",
  ],
  Array [
    1,
    "@import url(https://fonts.googleapis.com/css?family=Noto+Sans+TC|Roboto);",
    "",
  ],
  Array [
    1,
    "@import url();
@import url('');
@import url(\\"\\");
@import '';
@import \\"\\";
@import \\"   \\";
@import \\"
\\";
@import url();
@import url('');
@import url(\\"\\");
@import ;
@import foo-bar;
@import-normalize;
@import url('http://') :root {}

.class {
  a: b c d;
}

.foo {
  @import 'path.css';
}
",
    "",
  ],
]
`;

exports[`import option true: module 1`] = `
"exports = module.exports = require(\\"../../../lib/runtime/api.js\\")(false);
// imports
exports.i(require(\\"-!../../../index.js??ref--4-0!./test.css\\"), \\"\\");
exports.i(require(\\"-!../../../index.js??ref--4-0!./test.css\\"), \\"screen and print\\");
exports.i(require(\\"-!../../../index.js??ref--4-0!./test-media.css\\"), \\"screen and print\\");
exports.i(require(\\"-!../../../index.js??ref--4-0!./test-other.css\\"), \\"(min-width: 100px)\\");
exports.push([module.id, \\"@import url(http://example.com/style.css);\\", \\"\\"]);
exports.push([module.id, \\"@import url(http://example.com/style.css#hash);\\", \\"\\"]);
exports.push([module.id, \\"@import url(http://example.com/style.css?#hash);\\", \\"\\"]);
exports.push([module.id, \\"@import url(http://example.com/style.css?foo=bar#hash);\\", \\"\\"]);
exports.push([module.id, \\"@import url(http://example.com/other-style.css);\\", \\"screen and print\\"]);
exports.push([module.id, \\"@import url(//example.com/style.css);\\", \\"\\"]);
exports.i(require(\\"-!../../../index.js??ref--4-0!package/test.css\\"), \\"\\");
exports.i(require(\\"-!../../../index.js??ref--4-0!./query.css?foo=1&bar=1\\"), \\"\\");
exports.i(require(\\"-!../../../index.js??ref--4-0!./other-query.css?foo=1&bar=1#hash\\"), \\"\\");
exports.i(require(\\"-!../../../index.js??ref--4-0!./other-query.css?foo=1&bar=1#hash\\"), \\"screen and print\\");
exports.push([module.id, \\"@import url(https://fonts.googleapis.com/css?family=Roboto);\\", \\"\\"]);
exports.push([module.id, \\"@import url(https://fonts.googleapis.com/css?family=Noto+Sans+TC);\\", \\"\\"]);
exports.push([module.id, \\"@import url(https://fonts.googleapis.com/css?family=Noto+Sans+TC|Roboto);\\", \\"\\"]);

// module
exports.push([module.id, \\"@import url();\\\\n@import url('');\\\\n@import url(\\\\\\"\\\\\\");\\\\n@import '';\\\\n@import \\\\\\"\\\\\\";\\\\n@import \\\\\\"   \\\\\\";\\\\n@import \\\\\\"\\\\n\\\\\\";\\\\n@import url();\\\\n@import url('');\\\\n@import url(\\\\\\"\\\\\\");\\\\n@import ;\\\\n@import foo-bar;\\\\n@import-normalize;\\\\n@import url('http://') :root {}\\\\n\\\\n.class {\\\\n  a: b c d;\\\\n}\\\\n\\\\n.foo {\\\\n  @import 'path.css';\\\\n}\\\\n\\", \\"\\"]);

// exports
"
`;

exports[`import option true: warnings 1`] = `
Array [
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(12:1) Unable to find uri in '@import url()'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(13:1) Unable to find uri in '@import url('')'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(14:1) Unable to find uri in '@import url(\\"\\")'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(17:1) Unable to find uri in '@import '''",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(18:1) Unable to find uri in '@import \\"\\"'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(19:1) Unable to find uri in '@import \\"   \\"'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(20:1) Unable to find uri in '@import \\"
\\"'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(22:1) Unable to find uri in '@import url()'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(23:1) Unable to find uri in '@import url('')'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(24:1) Unable to find uri in '@import url(\\"\\")'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(40:1) Unable to find uri in '@import '",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(41:1) Unable to find uri in '@import foo-bar'",
  "ModuleWarning: Module Warning (from \`replaced original path\`):
Warning

(43:1) It looks like you didn't end your @import statement correctly. Child nodes are attached to it.",
]
`;
